### PR TITLE
fix pod with requiredDuringSchedulingIgnoredDuringExecution behave incorrectly after kubelet restart

### LINKED ISSUE #76322: pod with `requiredDuringSchedulingIgnoredDuringExecution` behave incorrectly after kubelet restart
**What happened**:
we have a pod with `spec.affinity.nodeAffinity.requiredDuringSchedulingIgnoredDuringExecution` and running peacefully at specific node, let's say node1, when node1's label changed and mismatched pod's constraint specified by node affinity, the pod will run continually as the `IgnoredDuringExecution` semantic.  However if we restart the kubelet agent, this pod will be killed with  `MatchNodeSelector`  status, and another pod will create by owner deployment to meet deployment's replicas num.

after digging into kubelet's code, I found when kubelet restart, it will get all pods belonging to this node from apiserver,  and execute `handler.HandlePodAdditions`  function, then kubelet will check if this pod is admitted; if not, reject it. In the checking process, `predicates.GeneralPredicates`  will fail to admit pod.

**What you expected to happen**:
After kubelet restart, don't kill pods and lets it running as before during execution even though this node doesn't match pod's affinity label.
the restart of kubelet shouldn't break `IgnoredDuringExecution` semantics.

**How to reproduce it (as minimally and precisely as possible)**:
1. add a label to specificy node, let's say 'node1' :
```yaml
kubectl label node node1 test=test
```
2. create a deployment  with `spec.affinity.nodeAffinity.requiredDuringSchedulingIgnoredDuringExecution` as  bellow:
```yaml
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: nginx
  namespace: default
  labels:
    resource_name: nginx
spec:
  selector:
    matchLabels:
      resource_name: nginx
  template:
    metadata:
      labels:
        resource_name: nginx 
    spec:
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: test
                operator: In
                values:
                - test
      containers:
      - image: nginx
        name: nginx
```
a nginx pod will create by this deployment.
3. remove node1's lable
```yaml
kubectl label node node1 test-
```
till now, everything is ok since  `requiredDuringSchedulingIgnoredDuringExecution` indicates that label changing should affect this pod.
```
[root@k8s-test-master01 test]# kubectl get pods
NAME                   READY   STATUS    RESTARTS   AGE
nginx-7cbbd59c-6xjmn   1/1     Running   0          9m32s
```

4. restart kubelet agent:
```
systemctl restart kubelet
```
After kubelet restart, this pod will be killed by kubelet as below,  another pod will recreate on another node.
```
[root@k8s-test-master01 test]# kubectl get pods
NAME                   READY   STATUS              RESTARTS   AGE
nginx-7cbbd59c-6xjmn   0/1     MatchNodeSelector   0          10m
nginx-7cbbd59c-w9t2c   0/1     Running             0          9s
```

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.0", GitCommit:"641856db18352033a0d96dbc99153fa3b27298e5", GitTreeState:"clean", BuildDate:"2019-03-25T15:53:57Z", GoVersion:"go1.12.1", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.0", GitCommit:"641856db18352033a0d96dbc99153fa3b27298e5", GitTreeState:"clean", BuildDate:"2019-03-25T15:45:25Z", GoVersion:"go1.12.1", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration:
bare metal
- OS (e.g: `cat /etc/os-release`):
CentOS Linux 7
- Kernel (e.g. `uname -a`):
Linux k8s-test-master01 3.10.0-862.el7.x86_64 #1 SMP Fri Apr 20 16:44:24 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by gaorong**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
After kubelet restarts, kubelet will get all pods belonging to him from apiserver, and execute `canAdmitPod` function to check if this pod is admitted; if not, reject it. 
For a pod with `requiredDuringSchedulingIgnoredDuringExecution` spec,  if it's previously located node's label has changed and doesn't meet the pod's affinity constraint, `canAdmitPod`, specifically  `predicates.GeneralPredicates`,  will fail to admit pod in the checking process, IMHO, this behavire violate the `requiredDuringSchedulingIgnoredDuringExecution` semantic. basically, the restart of kubelet shouldn't make an effect on an existing running pod. this PR fix this issue by ignoring node affinity check for pods at running status.

**Which issue(s) this PR fixes**:

Fixes #76322 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/scheduler/algorithm/predicates/predicates.go [hunk 1]
```diff
@@ -881,11 +881,15 @@ func podMatchesNodeSelectorAndAffinityTerms(pod *v1.Pod, node *v1.Node) bool {
 
 		// Match node selector for requiredDuringSchedulingIgnoredDuringExecution.
 		if nodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution != nil {
-			nodeSelectorTerms := nodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms
-			klog.V(10).Infof("Match for RequiredDuringSchedulingIgnoredDuringExecution node selector terms %+v", nodeSelectorTerms)
-			nodeAffinityMatches = nodeAffinityMatches && nodeMatchesNodeSelectorTerms(node, nodeSelectorTerms)
+			// If pod.Status.Phase is running, which means that this pod was previously scheduled and started running,
+			// the reason running this check is that kubelet is restarted and reruns canAdmitPod method,
+			// in this case, we will ignore NodeAffinity check to follow the *IgnoredDuringExectution* semantic.
+			if pod.Status.Phase != v1.PodRunning {
```
Comment (bsalamat) [inline pkg/scheduler/algorithm/predicates/predicates.go]: How could a pod which is being scheduled be in "running" state? Could you please describe the bug that you are trying to fix?

Comment (hex108) [reply pkg/scheduler/algorithm/predicates/predicates.go]: @bsalamat Kubelet also use scheduler's predicates for admitting pods.

Comment (hex108) [inline pkg/scheduler/algorithm/predicates/predicates.go]: How about `if nodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution != nil && len(pod.Spec.NodeName) != 0 {`?

File: pkg/scheduler/algorithm/predicates/predicates.go [hunk 2]
```diff
@@ -881,11 +881,15 @@ func podMatchesNodeSelectorAndAffinityTerms(pod *v1.Pod, node *v1.Node) bool {
 
 		// Match node selector for requiredDuringSchedulingIgnoredDuringExecution.
 		if nodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution != nil {
-			nodeSelectorTerms := nodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms
-			klog.V(10).Infof("Match for RequiredDuringSchedulingIgnoredDuringExecution node selector terms %+v", nodeSelectorTerms)
-			nodeAffinityMatches = nodeAffinityMatches && nodeMatchesNodeSelectorTerms(node, nodeSelectorTerms)
+			// If pod.Status.Phase is running, which means that this pod was previously scheduled and started running,
+			// the reason running this check is that kubelet is restarted and reruns canAdmitPod method,
+			// in this case, we will ignore NodeAffinity check to follow the *IgnoredDuringExectution* semantic.
```
[the remark this answers is not in the document]
Comment (gaorong) [reply pkg/scheduler/algorithm/predicates/predicates.go]: sure, how about this time?

File: pkg/scheduler/algorithm/predicates/predicates.go (hunk 1 above)
Comment (gaorong) [reply pkg/scheduler/algorithm/predicates/predicates.go]: @bsalamat  sorry for the confusing,  I've updated the description of this PR, You can also take a look at the corresponding issue.

Comment (gaorong) [reply pkg/scheduler/algorithm/predicates/predicates.go]: if someone explicitly assign a pod to a node  by filling in `pod.Spec.NodeName`, in this scenario, the kubelet will allow pod althought this node's label doesn't meet pod's affinity constraint.

Comment (gaorong) [reply pkg/scheduler/algorithm/predicates/predicates.go]: After revisiting this issue, I changed as you comment which seems to be more clear. 

> if someone explicitly assign a pod to a node by filling in `pod.Spec.NodeName`, in this scenario, the kubelet will allow pod although this node's label doesn't meet pod's affinity constraint.
This is a common issue and is not the goal of this PR.

**Comment by gaorong** [comment]
After revisiting the corresponding issue, I think maybe we should not change logic in scheduler to fix this, to some extent,  the root cause is kubelet does not support `restore` action (for now kubelet still treat `restore` as `add`: [see here](https://github.com/kubernetes/kubernetes/blob/f18833ee582d9f3bbe57e5f7c97f933c455640dc/pkg/kubelet/kubelet.go#L1919) and only apply to those pod with specific annotaion)
For a pod already be running state, kubelet should execute restore-related logical and shouldn't use scheduler's predicates to check if this pod can be admitted again.

But this PR would be okay for a short-term solution before kubelet support `restore` action.

feel free to close this if this is not wanted

**Pull request closed without merging**